explainer.md

```diff
@@ -1,624 +1,397 @@
+
+## Signaling the end of a page transition
+
+At the start of the transition, the browser could gather all the `Animation`s active on the stage, and assume the animation is complete once all animations finish.
+
+In addition, the JS API could include a way to provide a promise which keeps the transition active, allowing for animations driven some other way, such as `requestAnimationFrame`.
+
+## Example
+
+Using the sketches above, here's how the example Page-A to Page-B transition could be done:
+
+https://user-images.githubusercontent.com/93594/141100217-ba1fa157-cd79-4a9d-b3b4-67484d3c7dbf.mp4
+
+In Page-A:
+
+```css
+.header {
+  page-transition-tag: header;
+  page-transition-capture: container-and-child;
+}
+
+.share-button {
+  page-transition-tag: share-button;
+}
+```
+
+In Page-B:
+
+```css
+.header {
+  page-transition-tag: header;
+  page-transition-capture: container-and-child;
+}
+
+.share-button {
+  page-transition-tag: share-button;
+}
+
+/* Slide the roots from right to left */
+@keyframes slide-left {
+  to { transform: translateX(-100%); }
+}
+
+::page-transition-root-outgoing {
+  animation-name: slide-left;
+}
+
+::page-transition-root-incoming {
+  left: 100%;
+  animation-name: slide-left;
+}
+
+/* Prevent the header content from stretching */
+::page-transition-texture-outgoing(share-button),
+::page-transition-texture-incoming(share-button) {
+  object-fit: cover;
+}
+```
+
+# Single-Page-App API
+
+The mechanism for cross-document transitions and SPA transitions involves the same phases, so an SPA API will expose those parts in the same page.
+
+Half-baked sketch:
 
 ```js
-requestAnimationFrame(() => {
-  let pendingTransition = document.documentTransition.getPendingTransition();
-  if (pendingTransition.getData().version !== 123)
-    return;
-
-  // |offeredTransitionElements| provides a list of objects to access state
-  // saved from the old page.
-  let oldHeader = pendingTransition.offeredTransitionItems.get("header-id");
-  if (oldHeader) {
-    // Add sharedid attribute to elements animated in the new DOM.
-    document.querySelector(".header").sharedid="header-id";
-
-    // Query the style information saved from the old page.
-    let oldHeaderStyle = oldHeader.getContainerComputedStyle();
-
-    // The pseudo elements for each shared element are associated with the root element.
-    // The existing [pseudoElement](https://drafts.csswg.org/web-animations-1/#dom-keyframeeffect-pseudoelement) option can be used
-    // to target them with Web Animations API.
-    document.documentElement.animate(
-      [{ width: oldHeaderStyle.width,
-        height: oldHeaderStyle.height,
-        transform: oldHeaderStyle.transform }],
-      { duration: 1000,
-        pseudoElement: '::shared-container(header-id)' });
-  }
+document.documentTransition.prepare(async (transition) => {
+  // …Identify State-A elements to capture, then:
```

Comment:
I'm assuming the async callback is dispatched *after* the prepare phase finishes and the next scene can be asynchronously loaded. So identifying state-A elements can't be in this callback.

Reply:
In this sketch, it's in state-a before `await updatePageStateSomehow()` is called. I went with this patter for some of the same reasons as https://github.com/WICG/shared-element-transitions/issues/24, but it's just a quick sketch and I haven't really thought it through.